File: public/python/CDEEPSO.py
from techno_ka import *
from structure import *
from random import random, uniform
from math import sqrt, inf
import numpy as np
from time import time

init_t = time()

## Problem definition
C=0.65#price of electricity
W=0.14#loss of load probability#
K=0.9#renewable energy factor#
nvars=1#only one system

LB=[10, 1, 1, 1] # Lower bound of problem
UB=[150, 3, 30, 8] # Upper bound of problem
# LB=[130, 2, 15, 150] # Lower bound of problem
# UB=[130, 2, 15, 150] # Upper bound of problem

max_it=10#100# Maximum number of iterations
NPOP=10#20# Number of population
# Determine the maximum step for velocity
for d in range(4):
    if (LB[d]>-1e20) and (UB[d]<1e20):
        velmax = (UB[d]-LB[d])/NPOP
    else:
        velmax = inf

## algorithm initial parameters
phi1=2.05
phi2=2.05
phi=phi1+phi2
chi=2/(phi-2+sqrt(phi**2-4*phi))

w1=chi                                # Inertia weight
c1=chi*phi1                           # Personal learning coefficient
c2=chi*phi2                           # Global learning coefficient

## Initialization

particle=[]
for i in range(NPOP):
    particle.append(empty_particle([],[],[],best([],[])))
    
globalbest.cost=inf
globalbest.position=[]

for i in range(NPOP):
    cc=1#a value for cost
    ww=0.3# a value for lose of load probability#
    kkk=2#renewable energy factor# 2
    ff=0

    while ww>=W or kkk>=K:# LOLP/REfactor
        particle[i].position.extend(np.random.uniform(LB, UB).tolist())
        # particle[i].position.append (uniform(10,150))#pv kW # Rated PV power
        # particle[i].position.append (uniform(1,3))#autonomy days #
        # particle[i].position.append (uniform(1,30))#number of houses
        # particle[i].position.append (uniform(1,8))#number of wind turbine
        
        for g in range(4):
            particle[i].velocity.append (random())
                
        #----convert------------
        p_npv=particle[i].position[0]
        ad=particle[i].position[1]
        houses=round(particle[i].position[2])
        nwt=round(particle[i].position[3])
        #-----------------------
        [LPSP,price_electricity,renewable_factor,b,ali, ali2] = techno_ka(houses,p_npv,ad,nwt)
        ff = ff+1
        ww = LPSP
        kkk= renewable_factor

    particle[i].cost=price_electricity
    particle[i].best.position=particle[i].position
    particle[i].best.cost=particle[i].cost
    if particle[i].best.cost<globalbest.cost:
        globalbest=particle[i].best
        
Fminn=np.zeros(max_it) #EN MATLAB ES UNA COLUMNA
## algorithm main loop
for u in range(max_it):
    vv=0
    for i in range(NPOP):
        cc=1#a value for cost
        LPSP=0.3# a value for lose of load probability#
        renewable_factor=2#
        bb=0
        while (LPSP)>=W or (renewable_factor)>=K:
            for y in range(4):
                
                mut = random()
                particle[i].velocity[y] =( (w1*mut)*particle[i].velocity[y]+(c1*mut)*random()
                    *(particle[i].best.position[y]-particle[i].position[y])
                    +(c2*mut)*random()*(globalbest.position[y]-particle[i].position[y]) )
                
                particle[i].position[y]=particle[i].position[y]+particle[i].velocity[y]
                # particle[i].position[y]=min(max(particle[i].position[y], LB[y]), UB[y])
                particle[i].position[y]=np.clip(particle[i].position[y], LB[y], UB[y])
                          

            oo=0
            p_npv = round(particle[i].position[0])
            ad = round(particle[i].position[1])
            houses = round(particle[i].position[2])
            nwt = round(particle[i].position[3])
           
            [LPSP,price_electricity,renewable_factor,b,ali, ali2] = techno_ka(houses,p_npv,ad,nwt)
            bb=bb+1

        #-----------------------
        [LPSP,price_electricity,renewable_factor,b,ali, ali2] = techno_ka(houses,p_npv,ad,nwt)
        particle[i].cost = price_electricity
        rnwfct = renewable_factor 
        vv = vv+1
        if particle[i].cost < particle[i].best.cost:
            particle[i].best.cost = particle[i].cost
            particle[i].best.position = particle[i].position
            if particle[i].best.cost < globalbest.cost: #& rnwfct<rnwfct_best
                globalbest = particle[i].best
                rnwfct_best = rnwfct #

    
    Fminn[u] = globalbest.cost
    Xmin = globalbest.position
    p_npv = round(globalbest.position[0])
    ad = round(globalbest.position[1])
    houses = round(globalbest.position[2])
    nwt = round(globalbest.position[3])
    print(str.format("Iteration {0}, Best cost = {1}",u,Fminn[u]))
    print(str.format("Best solution p_npv = {0}, ad = {1}, houses = {2}, nwt = {3}", p_npv, ad, houses, nwt))

[LPSP,price_electricity,renewable_factor,b,ali, ali2] = techno_ka(houses,p_npv,ad,nwt)
print(str.format("LOLP {0}, $/KWh = {1}, %RES = {2}", LPSP, price_electricity, renewable_factor))

print("Total time {}".format(time() - init_t))

File: public/js/CDEEPSO.js
import { techno_ka } from './techno_ka.js';
import { emptyparticle, best, globalbest } from './structure.js';

export default async function CDEEPSO(iterations, steps, shouldPause, fixedValues, startIteration = 0) {
    const init_t = Date.now() / 1000;

    let runningStatus = document.getElementById("running_sim_status"); // apenas para avisar em qual iteração está

    // Problem definition
    const C = 0.65; // price of electricity
    const W = 0.14; // loss of load probability
    const K = 0.9;  // renewable energy factor
    const nvars = 1; // only one system

    const LB = [10, 1, 1, 1];  // Lower bound (p_npv, ad, houses, nwt)
    const UB = [150, 3, 30, 8]; // Upper bound

    const max_it = 10;
    const NPOP = 10;
    let velmax = [];

    for (let d = 0; d < 4; d++) {
        if (LB[d] > -1e20 && UB[d] < 1e20) {
            velmax[d] = (UB[d] - LB[d]) / NPOP;
        } else {
            velmax[d] = Infinity;
        }
    }

    // Algorithm initial parameters
    const phi1 = 2.05;
    const phi2 = 2.05;
    const phi = phi1 + phi2;
    const chi = 2 / (phi - 2 + Math.sqrt(Math.pow(phi, 2) - 4 * phi));

    const w1 = chi;               // Inertia weight
    const c1 = chi * phi1;        // Personal learning coefficient
    const c2 = chi * phi2;        // Global learning coefficient

    // Initialization
    let particle = [];

    for (let i = 0; i < NPOP; i++) {
        particle.push(new emptyparticle([], [], [], new best([], [])));
    }

    let global_best = new globalbest(Infinity, []);

    for (let i = 0; i < NPOP; i++) {
        let cc = 1;
        let ww = 0.3;
        let kkk = 2;
        let ff = 0;

        while (ww >= W || kkk >= K) {
            particle[i].position = LB.map((lb, index) => lb + Math.random() * (UB[index] - lb));
            particle[i].velocity = Array(4).fill(0).map(() => Math.random());

            let p_npv = particle[i].position[0];
            let ad = particle[i].position[1];
            let houses = Math.round(particle[i].position[2]);
            let nwt = Math.round(particle[i].position[3]);

            var [LPSP, price_electricity, renewable_factor, b, ali, ali2] = await techno_ka(houses, p_npv, ad, nwt, steps);
            ff += 1;
            ww = LPSP;
            kkk = renewable_factor;
        }
        
        particle[i].cost = price_electricity;
        particle[i].best.position = particle[i].position;
        particle[i].best.cost = particle[i].cost;

        if (particle[i].best.cost < global_best.cost) {
            global_best = particle[i].best;
        }
    }

    // Algorithm main loop
    let Fminn = Array(iterations).fill(0);

    let temp_renewable_factor, temp_LPSP, temp_price_electricity;

    let u = startIteration;
    for (u; u < iterations; u++) {
        let vv = 0;
        runningStatus.innerHTML = `Rodando iteração ${u+1}/${iterations}`;

        for (let i = 0; i < NPOP; i++) {
            if (shouldPause()) {
                return { pausedAtIteration: u }; // cancela a iteração atual se usuario pausou
            }

            let cc = 1;
            let LPSP = 0.3;
            let renewable_factor = 2;
            let bb = 0;

            while (LPSP >= W || renewable_factor >= K) {
                for (let y = 0; y < 4; y++) {
                    let mut = Math.random();
                    particle[i].velocity[y] = (
                        (w1 * mut) * particle[i].velocity[y] +
                        (c1 * mut) * Math.random() * (particle[i].best.position[y] - particle[i].position[y]) +
                        (c2 * mut) * Math.random() * (global_best.position[y] - particle[i].position[y])
                    );

                    particle[i].position[y] = particle[i].position[y] + particle[i].velocity[y];
                    particle[i].position[y] = Math.max(LB[y], Math.min(UB[y], particle[i].position[y]));    
                }

                let oo = 0;
                var p_npv = Math.round(particle[i].position[0]);
                var ad = Math.round(particle[i].position[1]);
                var houses = Math.round(particle[i].position[2]);
                var nwt = Math.round(particle[i].position[3]);

                // Sobrescreve se tiver fixado
                if (fixedValues.generation !== null) p_npv = fixedValues.generation;
                if (fixedValues.houses !== null) houses = fixedValues.houses;
                if (fixedValues.turbines !== null) nwt = fixedValues.turbines;

                [LPSP, price_electricity, renewable_factor, b, ali, ali2] = await techno_ka(houses, p_npv, ad, nwt, steps);
                bb += 1;
            }

            [LPSP, price_electricity, renewable_factor, b, ali, ali2] = await techno_ka(houses, p_npv, ad, nwt, steps);
            particle[i].cost = price_electricity;
            var rnwfct = renewable_factor;
            vv += 1;

            if (particle[i].cost < particle[i].best.cost) {
                particle[i].best.cost = particle[i].cost;
                particle[i].best.position = particle[i].position;

                if (particle[i].best.cost < global_best.cost) {
                    global_best = particle[i].best;
                    var rnwfct_best = rnwfct;
                }
            }

            temp_renewable_factor = renewable_factor;
            temp_LPSP = LPSP;
            temp_price_electricity = price_electricity;
        }
        
        Fminn[u] = global_best.cost;
        let Xmin = global_best.position;
        p_npv = fixedValues.generation !== null ? fixedValues.generation : Math.round(global_best.position[0]);
        ad = Math.round(global_best.position[1]);
        houses = fixedValues.houses !== null ? fixedValues.houses : Math.round(global_best.position[2]);
        nwt = fixedValues.turbines !== null ? fixedValues.turbines : Math.round(global_best.position[3]);

        console.log(`Iteration ${u}, Best cost = ${Fminn[u]}`);
        console.log(`Best solution p_npv = ${p_npv}, ad = ${ad}, houses = ${houses}, nwt = ${nwt}`);

        // mostrando valores temporários nos cards (exceto na primeira iteração)
        const metricas = document.querySelectorAll(".metrica-texto .valor");
        const valoresTemp = [
            (temp_renewable_factor * 100).toFixed(2).replace(".", ",") + "%",
            (temp_LPSP * 100).toFixed(2).replace(".", ",") + "%",
            "R$" + temp_price_electricity.toFixed(3).replace(".", ",") + "/kWh",
            houses,
            nwt,
            p_npv + " kWh"
        ];

        metricas.forEach((el, index) => {
            el.innerText = valoresTemp[index];
            el.classList.add("temp"); // aplicando opacidade
            el.style.display = "inline";
        });
    
    }

    runningStatus.innerHTML = ""; // apagando
    [LPSP, price_electricity, renewable_factor, b, ali, ali2] = await techno_ka(houses, p_npv, ad, nwt, steps);
    //console.log(`LOLP ${LPSP}, $/KWh = ${price_electricity}, %RES = ${renewable_factor}`);

    let simulationData = {
        "renewable_factor": renewable_factor,
        "price_electricity": price_electricity,
        "loss_load_probability": LPSP,
        "houses": houses,
        "num_wind_turbines": nwt,
        "max_generation": p_npv
    };

    // Salvando os dados gerados localmente
    localStorage.setItem("simulationData", JSON.stringify(simulationData));

    //console.log(`Total time ${((Date.now() / 1000) - init_t).toFixed(2)} seconds`);
}